### PR TITLE
Add ValidRelationships checks to BlocksProjectiles and Gate

### LINKED ISSUE #18717: Add ValidStances check to BlocksProjectiles
This would allow behaviour like Red Alert 2's WallTransparency, whereby friendly units can shoot through objects which enemies cannot. It makes player-built structures blocking projectiles a lot more of a viable feature.

eg. to make a wall block only enemy projectiles:

`BlocksProjectiles:
    ValidStances: Enemy`

### PR DESCRIPTION
**Pull request by abc013**
Closes #18717.

Testcase: Build a gate and walls in TS. Use disk throwers to test. The disks should fly through the gate and should be blocked by the walls. Enemies can fire through the walls, but not through the gate.

### CODE REVIEW

**Comment by abcdefg30** [comment]
This needs a rebase because of #18677 now. (Ideally you wait until #18678 is merged.)

**Comment by abc013** [comment]
Rebased. Sorry for the delay.

File: OpenRA.Mods.Common/Traits/BlocksProjectiles.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BlocksProjectiles.cs]: Style nit: Can we please write this as:
```csharp
var blockers = a.TraitsImplementing<IBlocksProjectiles>()
	.Where(t => t.IsTraitEnabled() && t.ValidRelationships.HasStance(a.Owner.RelationshipWith(owner)))
	.ToList();
```
?

**Review by pchote — DISMISSED** [review]
LGTM, but one minor style nit and needs a rebase. Can you also please remove the testcase?

File: OpenRA.Mods.Common/Traits/BlocksProjectiles.cs
Comment (abc013) [reply OpenRA.Mods.Common/Traits/BlocksProjectiles.cs]: Of course.

**Comment by abc013** [comment]
Rebased and removed the testcase. Sorry for the delay.

**Comment by teinarss** [comment]
Needs a repush to fix errors in the checks.

**Comment by abc013** [comment]
Done and rebased for good measure.

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ab1b72e948a1043a5ff56029a9084f285139a6a9...d8d567146e4e6d887c71e60446cad7ed183d8c8f

**Pull request merged**
